### PR TITLE
FastColoredTextBox + UI changes

### PR DESCRIPTION
**Pull request by paolosalvatori**
- Replaced TextBox controls used for reading and writing the payload of messages with FastColoredTextBox
 - Fixed an error in Service Bus Relay test control 
 - Refactored the user interface of controls and forms used to enter o display the payload of messages
 - Changed Azure icon in the title of all forms
 - Upgraded WindowsAzure.ServiceBus NuGet package to version 4.1.8
 - Upgraded Newtonsoft.Json NuGet package to version 11.0.0

Note: the dll containing the FastColoredTextBox control is a customization of the control available on github at https://github.com/PavelTorgashov/FastColoredTextBox.
For the momement, I just referenced the dll and added the dll in a folder called Library. I know that adding binaries to a project on GitHub is not a good practice,
but I cannot create a NuGet package for the customization of a library that is already available as a NuGet package. Thoughts?

### CODE REVIEW

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.1.0-PullRequest0236.248.build.430 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.248.build.430) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/912b545383 by @paolosalvatori)

**Comment by SeanFeldman** [comment]
:x: [Build ServiceBusExplorer 4.1.0-PullRequest0236.249.build.432 failed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.249.build.432) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/9f0567d2c7 by @paolosalvatori)

**Comment by tomkerkhove** [comment]
Thanks! Would you mind adding some screenshots please?
________________________________

**Comment by SeanFeldman** [comment]
Warning: the build is still failing. Will review tonight.

> Would you mind adding some screenshots please?

Pfff! Can't you read code @tomkerkhove ?! :troll:

File: src/ServiceBusExplorer/Controls/TestRelayControl.cs [hunk 1]
```diff
@@ -236,17 +246,32 @@ private void InitializeControls()
...
+                    txtMessageText.Text = messageText;
+                }
+
                 // Set Tooltips
                 toolTip.SetToolTip(txtMessageCount, MessageCountTooltip);
                 toolTip.SetToolTip(txtSendTaskCount, SendTaskCountTooltip);
 
                 splitContainer.SplitterWidth = 16;
-                headersDataGridView.Size = txtMessageText.Size;
+                //headersDataGridView.Size = txtMessageText.Size;
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Controls/TestRelayControl.cs]: If not needed, should it be removed?

File: src/ServiceBusExplorer/Forms/MessageForm.cs [hunk 2]
```diff
@@ -112,7 +113,22 @@ public MessageForm(BrokeredMessage brokeredMessage, ServiceBusHelper serviceBusH
 
             messagePropertyGrid.SelectedObject = brokeredMessage;
 
-            txtMessageText.Text = JsonSerializerHelper.Indent(XmlHelper.Indent(serviceBusHelper.GetMessageText(brokeredMessage, out _)));
+            var messageText = serviceBusHelper.GetMessageText(brokeredMessage, out _);
+
+            if (JsonSerializerHelper.IsJson(messageText))
+            {
+                txtMessageText.Language = Language.JSON;
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Forms/MessageForm.cs]: This block is repeating itself so many times.
Could to something like `LanguageDetector` that would take a text and return a tuple `(language, identedText)`.
That way no need to 
1. perform all of these `if`s everywhere
2. remember to accommodate all of places this code is used when another formatting is added

File: src/ServiceBusExplorer/Helpers/CustomPropertyDescriptor.cs [hunk 3]
```diff
@@ -163,27 +163,28 @@ public override string Category
...
-                {
-                    sResult = attr.Category;
-                }
-                if (string.IsNullOrWhiteSpace(sResult))
-                {
-                    sResult = base.Category;
-                }
-                return sResult.PadLeft(Category.Length + tabAppendCount, '\t');
+                return base.Category;
+                //var sResult = String.Empty;
```
Comment (SeanFeldman) [inline src/ServiceBusExplorer/Helpers/CustomPropertyDescriptor.cs]: image

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs [hunk 4]
```diff
@@ -951,8 +953,7 @@ private void GetMessages(bool peek, bool all, int count, IBrokeredMessageInspect
                 messagesSplitContainer.SplitterDistance = messagesSplitContainer.Width -
                                                           GrouperMessagePropertiesWith -
                                                           messagesSplitContainer.SplitterWidth;
-                messageListTextPropertiesSplitContainer.SplitterDistance = messageListTextPropertiesSplitContainer.Size.Height / 2 - 8;
-                messagesCustomPropertiesSplitContainer.SplitterDistance = messagesCustomPropertiesSplitContainer.Size.Width / 2 - 8;
+                messageMainSplitContainer.SplitterDistance = messageMainSplitContainer.Size.Height / 2 - 8;
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: What are these magical numbers?

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs [hunk 5]
```diff
@@ -2170,7 +2168,24 @@ private void messagesDataGridView_RowEnter(object sender, DataGridViewCellEventA
             {
                 brokeredMessage = bindingList[e.RowIndex];
                 messagePropertyGrid.SelectedObject = brokeredMessage;
-                txtMessageText.Text = XmlHelper.Indent(serviceBusHelper.GetMessageText(brokeredMessage, out _));
+
+                var messageText = serviceBusHelper.GetMessageText(brokeredMessage, out _);
+
+                if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Wrap in a method and pass textbox object given this is duplicated?

File: src/ServiceBusExplorer/Controls/ListenerControl.cs [hunk 6]
```diff
@@ -625,7 +625,23 @@ private void messagesDataGridView_RowEnter(object sender, DataGridViewCellEventA
             brokeredMessage = bindingList[e.RowIndex];
             messagePropertyGrid.SelectedObject = brokeredMessage;
 
-            txtMessageText.Text = XmlHelper.Indent(serviceBusHelper.GetMessageText(brokeredMessage, out _));
+            var messageText = serviceBusHelper.GetMessageText(brokeredMessage, out _);
+
+            if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/ListenerControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs [hunk 7]
```diff
@@ -2263,7 +2278,24 @@ private void deadletterDataGridView_RowEnter(object sender, DataGridViewCellEven
             {
                 deadletterMessage = bindingList[e.RowIndex];
                 deadletterPropertyGrid.SelectedObject = deadletterMessage;
-                txtDeadletterText.Text = XmlHelper.Indent(serviceBusHelper.GetMessageText(deadletterMessage, out _));
+
+                var messageText = serviceBusHelper.GetMessageText(deadletterMessage, out _);
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/PartitionListenerControl.cs [hunk 8]
```diff
@@ -647,7 +648,24 @@ private void eventDataDataGridView_RowEnter(object sender, DataGridViewCellEvent
             }
             currentEventData = bindingList[e.RowIndex];
             eventDataPropertyGrid.SelectedObject = currentEventData;
-            txtMessageText.Text = XmlHelper.Indent(serviceBusHelper.GetMessageText(currentEventData, out _));
+
+            var messageText = serviceBusHelper.GetMessageText(currentEventData.Clone(), out _);
+
+            if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/PartitionListenerControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestEventHubControl.cs [hunk 9]
```diff
@@ -283,10 +288,27 @@ private void InitializeControls()
                 propertiesDataGridView.ColumnHeadersDefaultCellStyle.BackColor = Color.FromArgb(215, 228, 242);
                 propertiesDataGridView.ColumnHeadersDefaultCellStyle.ForeColor = SystemColors.ControlText;
 
-                txtMessageText.Text = mainForm != null &&
+                var messageText = mainForm != null &&
                                       !string.IsNullOrWhiteSpace(mainForm.MessageText) ?
                                       XmlHelper.Indent(mainForm.MessageText) :
                                       DefaultMessageText;
+
+                if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestEventHubControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestEventHubControl.cs [hunk 10]
```diff
@@ -1366,6 +1380,49 @@ private void checkBoxSendBatch_CheckedChanged(object sender, EventArgs e)
         {
             txtSendBatchSize.Enabled = checkBoxSendBatch.Checked;
         }
+
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestEventHubControl.cs]: Can we use some more self-descriptive method names please?

File: src/ServiceBusExplorer/Controls/TestEventHubControl.cs [hunk 11]
```diff
@@ -1366,6 +1380,49 @@ private void checkBoxSendBatch_CheckedChanged(object sender, EventArgs e)
...
+                mainForm.MessageText = txtMessageText.Text;
+            }
+        }
+
+        private void cboMessageFormat_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            txtMessageText.ClearStylesBuffer();
+            txtMessageText.Range.ClearStyle(StyleIndex.All);
+
+            switch (cboMessageFormat.Text)
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestEventHubControl.cs]: Refactor to method given this is used in other places as well

File: src/ServiceBusExplorer/Controls/TestQueueControl.cs [hunk 12]
```diff
@@ -346,9 +349,26 @@ private void InitializeControls()
...
                 propertiesDataGridView.ColumnHeadersDefaultCellStyle.ForeColor = SystemColors.ControlText;
 
-                txtMessageText.Text = !string.IsNullOrWhiteSpace(mainForm?.MessageText) ?
-                                      JsonSerializerHelper.Indent(XmlHelper.Indent(mainForm.MessageText)) :
+                var messageText = mainForm != null &&
+                                      !string.IsNullOrWhiteSpace(mainForm.MessageText) ?
+                                      XmlHelper.Indent(mainForm.MessageText) :
                                       DefaultMessageText;
+
+                if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestQueueControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestQueueControl.cs [hunk 13]
```diff
@@ -2142,6 +2154,44 @@ protected override void Dispose(bool disposing)
...
+                                   cboMessageFormat.Size.Width + 1,
+                                   cboMessageFormat.Size.Height + 1);
+        }
+
+        private void cboMessageFormat_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            txtMessageText.ClearStylesBuffer();
+            txtMessageText.Range.ClearStyle(StyleIndex.All);
+
+            switch (cboMessageFormat.Text)
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestQueueControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestRelayControl.cs [hunk 14]
```diff
@@ -236,17 +246,32 @@ private void InitializeControls()
...
 
-                txtMessageText.Text = mainForm != null &&
-                                      !string.IsNullOrWhiteSpace(mainForm.RelayMessageText) ?
-                                      XmlHelper.Indent(mainForm.RelayMessageText) :
+                var messageText = mainForm != null &&
+                                      !string.IsNullOrWhiteSpace(mainForm.MessageText) ?
+                                      XmlHelper.Indent(mainForm.MessageText) :
                                       DefaultMessageText;
 
+                if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestRelayControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestRelayControl.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/ServiceBusExplorer/Controls/TestRelayControl.cs]: :+1:

File: src/ServiceBusExplorer/Controls/TestRelayControl.cs [hunk 15]
```diff
@@ -1217,6 +1234,44 @@ protected override void Dispose(bool disposing)
...
+                                   cboMessageFormat.Size.Width + 1,
+                                   cboMessageFormat.Size.Height + 1);
+        }
+
+        private void cboMessageFormat_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            txtMessageText.ClearStylesBuffer();
+            txtMessageText.Range.ClearStyle(StyleIndex.All);
+
+            switch (cboMessageFormat.Text)
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestRelayControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestTopicControl.cs [hunk 16]
```diff
@@ -346,9 +349,26 @@ private void InitializeControls()
...
                 propertiesDataGridView.ColumnHeadersDefaultCellStyle.ForeColor = SystemColors.ControlText;
 
-                txtMessageText.Text = !string.IsNullOrWhiteSpace(mainForm?.MessageText) ?
-                                      JsonSerializerHelper.Indent(XmlHelper.Indent(mainForm.MessageText)) :
+                var messageText = mainForm != null &&
+                                      !string.IsNullOrWhiteSpace(mainForm.MessageText) ?
+                                      XmlHelper.Indent(mainForm.MessageText) :
                                       DefaultMessageText;
+
+                if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestTopicControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Controls/TestTopicControl.cs [hunk 17]
```diff
@@ -2146,6 +2158,44 @@ protected override void Dispose(bool disposing)
...
+                                   cboMessageFormat.Size.Width + 1,
+                                   cboMessageFormat.Size.Height + 1);
+        }
+
+        private void cboMessageFormat_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            txtMessageText.ClearStylesBuffer();
+            txtMessageText.Range.ClearStyle(StyleIndex.All);
+
+            switch (cboMessageFormat.Text)
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Controls/TestTopicControl.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Forms/EventDataForm.cs [hunk 18]
```diff
@@ -81,7 +82,22 @@ public EventDataForm(EventData eventData, ServiceBusHelper serviceBusHelper, Wri
             InitializeComponent();
 
             messagePropertyGrid.SelectedObject = eventData;
-            txtMessageText.Text = XmlHelper.Indent(serviceBusHelper.GetMessageText(eventData, out _));
+            var messageText = serviceBusHelper.GetMessageText(eventData, out _);
+
+            if (JsonSerializerHelper.IsJson(messageText))
```
Comment (tomkerkhove) [inline src/ServiceBusExplorer/Forms/EventDataForm.cs]: Can be reused here as well then

File: src/ServiceBusExplorer/Forms/MessageForm.cs (hunk 2 above)
Comment (tomkerkhove) [reply src/ServiceBusExplorer/Forms/MessageForm.cs]: 👍

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 4 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: " / 2" is "half". Not magical.
"- 8" is magical 🙂

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 5 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Related to https://github.com/paolosalvatori/ServiceBusExplorer/pull/236#discussion_r202419305

File: src/ServiceBusExplorer/Controls/TestRelayControl.cs (hunk 1 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Controls/TestRelayControl.cs]: Sure, feel free to remove any code that is commented out.

File: src/ServiceBusExplorer/Forms/MessageForm.cs (hunk 2 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Forms/MessageForm.cs]: Yes, I had the same idea.

File: src/ServiceBusExplorer/Helpers/CustomPropertyDescriptor.cs (hunk 3 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Helpers/CustomPropertyDescriptor.cs]: I made a change to remove a recurring bug, then I forgot to remove the code below. Let's remove it.

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 4 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Don't make changes please, there's a reason why I had to introduce that code, that is, even if I set the right value at design time, it looks like it gets lost. It's probably a bug in the splitter control.

File: src/ServiceBusExplorer/Controls/ListenerControl.cs (hunk 6 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Controls/ListenerControl.cs]: There are some places where I still use TextBox controls and not FastColoredTextBox, but I will check.

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 7 above)
Comment (paolosalvatori) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: Yep, let me fix that.

**Comment by paolosalvatori** [comment]
I fixed all the issues in the current branch, please review the latest version

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.1.0-PullRequest0236.250.build.436 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.250.build.436) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/2448c555da by @paolosalvatori)

**Comment by tomkerkhove** [comment]
There is some small duplication left:
```csharp
switch (cboMessageFormat.Text)
{
    case "JSON":
        txtMessageText.Language = Language.JSON;
        break;
    case "XML":
        txtMessageText.Language = Language.HTML;
        break;
    default:
        txtMessageText.Language = Language.Custom;
        break;
}
```

Besides that, looks good 👍

File: src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs (hunk 4 above)
Comment (SeanFeldman) [reply src/ServiceBusExplorer/Controls/HandleSubscriptionControl.cs]: In such case making an inline comment could be a good idea

**Comment by SeanFeldman** [comment]
Minor tweaks to the `LanguageDetector` and good to go

**Comment by paolosalvatori** [comment]
I think we are good now. Please review changes, thanks :)

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.1.0-PullRequest0236.251.build.438 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.251.build.438) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/48a42e56be by @paolosalvatori)

**Comment by SeanFeldman** [comment]
Changes look good, but what's with the duplication app.config? 🙂

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.1.0-PullRequest0236.252.build.439 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.252.build.439) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/da884deb6c by @paolosalvatori)

**Comment by SeanFeldman** [comment]
:white_check_mark: [Build ServiceBusExplorer 4.1.0-PullRequest0236.253.build.442 completed](https://ci.appveyor.com/project/seanfeldman/servicebusexplorer/build/4.1.0-PullRequest0236.253.build.442) (commit https://github.com/paolosalvatori/ServiceBusExplorer/commit/ff7938ef78 by @paolosalvatori)

**Comment by paolosalvatori** [comment]
I forgot to remove my connection strings from the app.config, then I fixed the issue :) @SeanFeldman, @tomkerkhove, @ErikMogensen let me know if I can proceed and merge changes, thanks! ;)

**Comment by SeanFeldman** [comment]
👍 to merge

I'll sound like a broken record, but we REALLY shouldn't keep any secrets
in app.config (another issue is opened) 🙂

**Pull request merged**